### PR TITLE
add backup subcommand

### PR DESCRIPTION
**Pull request by MaximilianMeister**
* create
* download
* delete
* list

### CODE REVIEW

File: lib/crowbar/client/app/backup.rb [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    module App
+      class Backup < Base
+        desc "list",
+          "List existing backups"
+
+        long_desc <<-LONGDESC
+          `list` will print out a list of existing backups on the admin
+          node. You can display the list in different output
+          formats and you can filter the list by any search criteria.
+          With --format <format> option you can choose an output format
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Add an empty line, see it at https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/app/barclamp.rb#L28

File: lib/crowbar/client/app/backup.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+          with the available options table, json or plain. You can also
+          use the shortcut options --table, --json or --plain.
+          With --filter <filter> option you can limit the result of
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Add an empty line, see it at https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/app/barclamp.rb#L32

File: lib/crowbar/client/app/backup.rb [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+          type: :string,
+          default: nil,
+          banner: "<filter>",
+          desc: "Filter by criteria, display only data that contains filter"
+
+        def list
+          Command::Backup::List.new(
+            *command_params
+          ).execute
+        rescue SimpleCatchableError => e
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Wrong exception catching, e.g. https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/app/barclamp.rb#L72-L73

File: lib/crowbar/client/app/backup.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+          `create NAME` will create a new backup on the Administration Server.
+        LONGDESC
+
+        def create(name)
+          Command::Backup::Create.new(
+            *command_params(
+              filename: name
+            )
+          ).execute
+        rescue SimpleCatchableError => e
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Wrong exception catching, e.g. https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/app/barclamp.rb#L72-L73

File: lib/crowbar/client/app/backup.rb [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        LONGDESC
+
+        def download(name, created_at)
+          Command::Backup::Download.new(
+            *command_params(
+              name: name,
+              created_at: created_at
+            )
+          ).execute
+        rescue SimpleCatchableError => e
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Wrong exception catching, e.g. https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/app/barclamp.rb#L72-L73

File: lib/crowbar/client/app/backup.rb [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        end
+
+        desc "upload",
+          "Upload a backup to the Administration Server"
+
+        long_desc <<-LONGDESC
+          `upload FILE` will upload a backup to the Administration Server.
+        LONGDESC
+
+        def upload(fileupload)
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Name it just `file`

File: lib/crowbar/client/app/backup.rb [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+          Command::Backup::Upload.new(
+            *command_params(
+              fileupload: fileupload
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Name it `file: file`

File: lib/crowbar/client/app/backup.rb [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            )
+          ).execute
+        rescue SimpleCatchableError => e
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Wrong exception catching, e.g. https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/app/barclamp.rb#L72-L73

File: lib/crowbar/client/app/backup.rb [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        LONGDESC
+
+        def delete(name, created_at)
+          Command::Backup::Delete.new(
+            *command_params(
+              name: name,
+              created_at: created_at
+            )
+          ).execute
+        rescue SimpleCatchableError => e
```
Comment (tboerger) [inline lib/crowbar/client/app/backup.rb]: Wrong exception catching, e.g. https://github.com/crowbar/crowbar-client/blob/master/lib/crowbar/client/app/barclamp.rb#L72-L73

File: lib/crowbar/client/command/backup/create.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+            @request ||= Request::Backup::Create.new(
+              args
+            )
+          end
+
+          def execute
+            request.process do |request|
+              case request.code
+              when 200
+                say "Created Backup"
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/create.rb]: `Successfully created a backup`

File: lib/crowbar/client/command/backup/delete.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+            @request ||= Request::Backup::Delete.new(
+              args
+            )
+          end
+
+          def execute
+            request.process do |request|
+              case request.code
+              when 200
+                say "Deleted Backup"
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/delete.rb]: `Successfully deleted the backup`

File: lib/crowbar/client/command/backup/download.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+              args
+            )
+          end
+
+          def execute
+            request.process do |request|
+              case request.code
+              when 200
+                download(request.body)
+                say "Downloaded Backup"
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/download.rb]: `Successfully downloaded backup to #{path}`

File: lib/crowbar/client/command/backup/download.rb [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            @request ||= Request::Backup::Download.new(
+              args
+            )
+          end
+
+          def execute
+            request.process do |request|
+              case request.code
+              when 200
+                download(request.body)
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/download.rb]: `write(request.body)`

File: lib/crowbar/client/command/backup/download.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+                say "Downloaded Backup"
+              else
+                err request.parsed_response["error"]
+              end
+            end
+          end
+
+          protected
+
+          def download(body)
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/download.rb]: `def write(body)`

File: lib/crowbar/client/command/backup/download.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+            File.open("./#{args.name}-#{args.created_at}.tar.gz", "wb") do |f|
+              f.binmode
+              f.write body
+            end
+          end
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/download.rb]:
```ruby
def path
  "#{args.name}-#{args.created_at}.tar.gz"
end
```

File: lib/crowbar/client/command/backup/download.rb [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+              else
+                err request.parsed_response["error"]
+              end
+            end
+          end
+
+          protected
+
+          def download(body)
+            File.open("./#{args.name}-#{args.created_at}.tar.gz", "wb") do |f|
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/download.rb]: `path.open("wb") do |f|`

File: lib/crowbar/client/command/backup/list.rb [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+                  format: provide_format,
+                  headings: headings,
+                  values: Filter::Hash.new(
+                    filter: provide_filter,
+                    values: content_from(request)
+                  ).result
+                )
+
+                if formatter.empty?
+                  err "No Backups"
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/list.rb]: `No backups`

File: lib/crowbar/client/command/backup/upload.rb [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+              args
+            )
+          end
+
+          def execute
+            request.process do |request|
+              case request.code
+              when 200
+                upload
+                say "Uploaded Backup"
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/upload.rb]: `Successfully uploaded backup`

File: lib/crowbar/client/command/backup/upload.rb [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+            @request ||= Request::Backup::Upload.new(
+              args
+            )
+          end
+
+          def execute
+            request.process do |request|
+              case request.code
+              when 200
+                upload
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/upload.rb]: Return boolean result and display message related to that.

File: lib/crowbar/client/command/backup/download.rb (hunk 13 above)
Comment (tboerger) [inline lib/crowbar/client/command/backup/download.rb]: Return boolean result and display message related to that.

**Comment by tboerger** [comment]
Please remove the upload method from this pull request. We need to wait for an approval of the httmultiparty rubygem by legal before we can integrate it. Please create another pull request based on the changes from https://github.com/crowbar/crowbar-client/pull/48.

File: lib/crowbar/client/command/backup/download.rb [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+                end
+              else
+                err request.parsed_response["error"]
+              end
+            end
+          end
+
+          protected
+
+          def write(body)
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/download.rb]:
```ruby
def write(body)
  path.open("wb") do |f|
    f.binmode
    f.write body

    true
  end
rescue
  false
end
```

Comment (MaximilianMeister) [inline lib/crowbar/client/command/backup/download.rb]: fixed

**Comment by rsalevsky** [comment]
+1

File: lib/crowbar/client/command/backup/list.rb [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+          end
+
+          protected
+
+          def headings
+            ["Name", "Created", "Filename", "Path"]
+          end
+
+          def content_from(request)
+            [].tap do |row|
```
Comment (tboerger) [inline lib/crowbar/client/command/backup/list.rb]: Maybe replace it with the following?

```ruby
request.parsed_response.map do |row|
  row.slice(
    "name",
    "created_at",
    "filename",
    "path"
  )
end
```

That's more secure for future serverside field additions.

File: lib/crowbar/client/request/backup/download.rb [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+# limitations under the License.
+#
+
+module Crowbar
+  module Client
+    module Request
+      module Backup
+        class Download < Base
+          def headers
+            {
```
Comment (tboerger) [inline lib/crowbar/client/request/backup/download.rb]:
```ruby
super.easy_merge!(
  "Content-Type" => "application/octet-stream",
  "Accept" => "application/octet-stream"
)
```

Doing it the same as for the content to be able to extend the base in the future.

**Comment by tboerger** [comment]
:+1:

**Pull request merged**